### PR TITLE
[Inductor] Added aten.uniform_ decomp

### LINKED ISSUE #90815: [Inductor] decompose aten.uniform_
### 🚀 The feature, motivation and pitch

Currently `aten.uniform_` fallbacks to `FallbackKernel: torch.ops.aten.uniform_.default`. 
```python
import torch

def foo(x):
    return x.uniform_(-1, 1)

opt_foo = torch.compile(foo)
print(opt_foo(torch.randn(10)))
```
`aten.uniform_` is decomposable to something like `(high - low) * torch.rand(x) + low`. 

### Correctness
```
X ~ U(0, 1) 
[0,1)

(θ_high - θ_low) * X ~ U(0, (θ_high - θ_low))
[0, θ_high - θ_low)

(θ_high - θ_low) * X + θ_low ~ U(θ_low, θ_high)
[0 + θ_low, θ_high - θ_low + θ_low)
=[θ_low, θ_high)
```

 

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh

### PR DESCRIPTION
**Pull request by min-jean-cho**
Fixes #90815

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Comment by min-jean-cho** [comment]
Hi @ashokei @jgong5, please help with review, thanks!

**Review by jgong5 — APPROVED** [review]

**Comment by min-jean-cho** [comment]
Hi @Chillee @jansel, could you please help review this PR? Thanks!

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -559,6 +559,11 @@ def bernoulli_(self, p=0.5):
     return self.copy_(torch.rand_like(self, dtype=torch.float32) < p)
 
 
+@register_extra_random_decomp([aten.uniform_])
+def uniform_(self, low=0, high=1):
+    return self.copy_((high - low) * torch.rand_like(self) + low)
```
Comment (lezcano) [inline torch/_inductor/decomposition.py:564]: This formula may overflow. Please consider using a formula that does not overflow regardless of the values of `high` and `low`.

**Review by lezcano — COMMENTED** [review]
Why is this an inductor-specfici decomposition and not a PrimTorch reference?

File: torch/_inductor/decomposition.py (hunk 1 above)
Comment (jansel) [reply torch/_inductor/decomposition.py:564]: Have a link?

**Review by jansel — APPROVED** [review]

File: torch/_inductor/decomposition.py (hunk 1 above)
Comment (lezcano) [reply torch/_inductor/decomposition.py:564]: xref. https://github.com/pytorch/pytorch/issues/90815#issuecomment-1354403724

Comment (min-jean-cho) [reply torch/_inductor/decomposition.py:561]: Hi all, I believe we have aligned on this, https://github.com/pytorch/pytorch/issues/90815#issuecomment-1356854711.

**Comment by min-jean-cho** [comment]
Hi all, is this good to land? Thanks!

**Comment by lezcano** [comment]
The main comment from https://github.com/pytorch/pytorch/pull/90869#pullrequestreview-1218776167 (i.e. why is this not in PrimTorch or in `decompositions.py`) has not been addressed yet.

**Comment by min-jean-cho** [comment]
Thanks @lezcano for the comments, 

> why is this not in PrimTorch 

I think as @ngimel  comments https://github.com/pytorch/torchdynamo/issues/327#issuecomment-1357115780, since `randn` is already a primitive op at lower level, including inductor-level, which makes `randn` suited for being used for decomp (e.g., `normal_`), if `rand` is also a primitive op at lower level, then I think `rand` is suited for being used for decomp (e.g., `uniform_`) as well.  

> or in decompositions.py

If `randn`, `rand` are typically primitive ops at low level, including but not unique to inductor, then I think it makes sense that these decomps don't have to be inductor-unique, and move under generic `_decomp/decompositions.py`

**Review by lezcano — APPROVED** [review]
CI failures are real and need fixing, but otherwise this is good to go.

**Review by jgong5 — APPROVED** [review]

**Comment by min-jean-cho** [comment]
Thanks @lezcano, looks like the CI failures are now fixed.

**Review by ngimel — APPROVED** [review]

**Review by lezcano — COMMENTED** [review]
I think some autograd changes snuck in in some rebase, they don't look relevant to the PR.

**Review by albanD — CHANGES_REQUESTED** [review]
Yes, the changes to gradcheck should definitely be reverted

File: torch/_decomp/decompositions.py [hunk 2]
```diff
@@ -1963,6 +1963,12 @@ def uniform(
     )
 
 
+@register_decomposition(aten.uniform_)
+def uniform_(self, low=0, high=1, generator=None):
+    assert generator is None
```
Comment (min-jean-cho) [inline torch/_decomp/decompositions.py:1968]: Looks like the assert fails on some crossref mode CIs. 

@jansel, may I know why the `generator` arg in decomp is just ignored here:
https://github.com/pytorch/pytorch/blob/f21997099087c6db85e119cd5ef97e2d8a134a85/torch/_decomp/decompositions.py#L1522-L1527

And the context behind
https://github.com/pytorch/pytorch/blob/f21997099087c6db85e119cd5ef97e2d8a134a85/torch/_prims/__init__.py#L2676-L2677
I wonder if it's OK to ignore the `generator` arg from the decomp side if the backend can still get the seed from the default generator (if the default generator is global).

Comment (jansel) [reply torch/_decomp/decompositions.py:1968]: I didn't write that code.  Though that decomp seems wrong if the user provides a non-default generator.  We should fix that.

Comment (ngimel) [reply torch/_decomp/decompositions.py:1968]: Yeah we should assert that generator is default or none, otherwise decomp is wrong. Currently inductor is not set up to respect non-default generator. We probably need to skip the test on crossref builds.

Comment (min-jean-cho) [reply torch/_decomp/decompositions.py:1968]: Thanks @jansel, @ngimel, understood.  decomp should respect the function args, so better not just ignore `generator` arg. 

Done, https://github.com/pytorch/pytorch/pull/91833.

**Review by albanD — APPROVED** [review]

**Comment by min-jean-cho** [comment]
> We probably need to skip the test on crossref builds.

Is this good to merge? If so, please help merge this, thanks!

**Comment by ngimel** [comment]
PR is approved, if CI is clean you can merge it using pytorchbot commands.

**Comment by min-jean-cho** [comment]
> PR is approved, if CI is clean you can merge it using pytorchbot commands.

Cool! Thank you all for the helpful comments and feedbacks throughout this PR.
